cpp/src/search/search_ordered.cu

```diff
@@ -62,54 +63,39 @@ std::unique_ptr<column> search_ordered(table_view const& haystack,
 
   // This utility will ensure all corresponding dictionary columns have matching keys.
   // It will return any new dictionary columns created as well as updated table_views.
-  auto const matched = dictionary::detail::match_dictionaries({haystack, needles}, stream);
-
-  // Prepare to flatten the structs column
-  auto const has_null_elements   = has_nested_nulls(haystack) or has_nested_nulls(needles);
-  auto const flatten_nullability = has_null_elements
-                                     ? structs::detail::column_nullability::FORCE
-                                     : structs::detail::column_nullability::MATCH_INCOMING;
-
-  // 0-table_view, 1-column_order, 2-null_precedence, 3-validity_columns
-  auto const t_flattened = structs::detail::flatten_nested_columns(
-    matched.second.front(), column_order, null_precedence, flatten_nullability);
-  auto const values_flattened =
-    structs::detail::flatten_nested_columns(matched.second.back(), {}, {}, flatten_nullability);
-
-  auto const t_d      = table_device_view::create(t_flattened, stream);
-  auto const values_d = table_device_view::create(values_flattened, stream);
-  auto const& lhs     = find_first ? *t_d : *values_d;
-  auto const& rhs     = find_first ? *values_d : *t_d;
-
-  auto const& column_order_flattened    = t_flattened.orders();
-  auto const& null_precedence_flattened = t_flattened.null_orders();
-  auto const column_order_dv = detail::make_device_uvector_async(column_order_flattened, stream);
-  auto const null_precedence_dv =
-    detail::make_device_uvector_async(null_precedence_flattened, stream);
-
-  auto const count_it = thrust::make_counting_iterator<size_type>(0);
-  auto const comp     = row_lexicographic_comparator(nullate::DYNAMIC{has_null_elements},
-                                                 lhs,
-                                                 rhs,
-                                                 column_order_dv.data(),
-                                                 null_precedence_dv.data());
+  auto const matched          = dictionary::detail::match_dictionaries({haystack, needles}, stream);
+  auto const matched_haystack = matched.second.front();
+  auto const matched_needles  = matched.second.back();
+
+  auto const& lhs       = find_first ? matched_haystack : matched_needles;
+  auto const& rhs       = find_first ? matched_needles : matched_haystack;
+  auto const comparator = cudf::experimental::row::lexicographic::two_table_comparator(
+    lhs, rhs, column_order, null_precedence, stream);
```

Comment:
```suggestion
    matched_haystack, matched_needles, column_order, null_precedence, stream);
```

Reply:
I believe the existing code here, along with the ternary statements above, are required for the strongly typed comparator to work correctly.

Reply:
No, they were required because we didn't have strong typed index.
`thrust::lower_bound` calls `comp(lhs_index, rhs_index)` while `thrust::upper_bound` call `comp(rhs_index, lhs_index)`. As such, we have to switch the operands `lhs` and `rhs` to have the correct comparison. Now with strong typed, we just remove that switch.

Reply:
Aha, thank you for the correction. I had forgotten my original reason for keeping the previous code's table flipping behavior. Until I made the `operator()` symmetric by using weak orderings (so we have `lhs < rhs` and `rhs < lhs`, instead of the asymmetric `not(lhs < rhs) == (rhs <= lhs)`), this wouldn't have been possible. It required both strong typing _and_ symmetric operators. Now that we have symmetry in that operator, it works as expected. Fixed in 09c5661421071c8e4f3dfd497761390af36c6835.